### PR TITLE
feat: Add Type Safety to Cognitive Steps and Processors [OPE-580]

### PR DESCRIPTION
**Pull request by greynewell**
- [x] Hoist type
- [x] Add type safety to `Processor`
- [x] Add type safety to `CogitiveStep`

## QUESTIONS

- [x] ~Should we also hoist the `MODEL_MAP` here so that both can be seen and/or edited in the same place? This would also let us import it for use in other places (like our docs)~

### CODE REVIEW

**Comment by greynewell** [comment]
Screenshot 2024-06-04 at 10 16 10 AM
Looks like the issue is because `processor` is undefined, which exits the switch control-flow.

**Review by neilsonnn — APPROVED** [review]

**Comment by tobowers** [comment]
the "model" field can also look like "tobowers/my-custom-model"

**Comment by greynewell** [comment]
> the "model" field can also look like "tobowers/my-custom-model"

I think I came up with a cool solution here!
```
type OrganizationSlug = string;
type CustomModelName = string;
export type CUSTOM_MODEL = `${OrganizationSlug}/${CustomModelName}`;

export type SupportedModel = typeof SUPPORTED_MODELS[number] | CUSTOM_MODEL;
```

File: packages/core/src/sharedTypes/supportedModels.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export const SUPPORTED_MODELS = ["fast", "quality", "vision", "gpt-4-turbo", "gpt-4o", "exp/OpenHermes-2p5-Mistral-7B", 
+"exp/Nous-Hermes-2-Mixtral-8x7B-SFT", "exp/Nous-Hermes-2-Mixtral-8x7B-DPO", "exp/claude-3-opus", "exp/claude-3-sonnet",
+"exp/claude-3-haiku", "exp/nous-hermes-2-mixtral-fp8", "exp/hermes-2-pro-mistral-7b", "exp/firefunction-v1", "exp/firellava-13b",
+"exp/mixtral-8x22b-instruct", "exp/llama-v3-70b-instruct"];
+
+type OrganizationSlug = string;
+type CustomModelName = string;
+export type CUSTOM_MODEL = `${OrganizationSlug}/${CustomModelName}`;
```
Comment (dooart) [inline packages/core/src/sharedTypes/supportedModels.ts:8]: i had no idea you could do this!

File: packages/core/src/sharedTypes/supportedModels.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export const SUPPORTED_MODELS = ["fast", "quality", "vision", "gpt-4-turbo", "gpt-4o", "exp/OpenHermes-2p5-Mistral-7B", 
+"exp/Nous-Hermes-2-Mixtral-8x7B-SFT", "exp/Nous-Hermes-2-Mixtral-8x7B-DPO", "exp/claude-3-opus", "exp/claude-3-sonnet",
+"exp/claude-3-haiku", "exp/nous-hermes-2-mixtral-fp8", "exp/hermes-2-pro-mistral-7b", "exp/firefunction-v1", "exp/firellava-13b",
```
Comment (dooart) [inline packages/core/src/sharedTypes/supportedModels.ts:3]: isn't this missing those new ones you introduced in platform?

"gpt-3.5-turbo-0125", "gpt-4-0125-preview", "gpt-4-vision-preview"

**Review by dooart — CHANGES_REQUESTED** [review]

File: packages/core/src/sharedTypes/supportedModels.ts (hunk 2 above)
Comment (greynewell) [reply packages/core/src/sharedTypes/supportedModels.ts:3]: @dooart oh, true!

I'll raise a PR later so that `platform` reads from this as well.

Comment (greynewell) [reply packages/core/src/sharedTypes/supportedModels.ts:3]: Fixed in next revision.

File: packages/core/src/sharedTypes/supportedModels.ts (hunk 1 above)
Comment (greynewell) [reply packages/core/src/sharedTypes/supportedModels.ts:8]: @dooart I love string template types! It does mean basically any input is valid as long as it has a `/`, but it's really nice for type suggestions specifically.

I am hesitant to get tricky with types for domain objects, but for surface APIs it lends expressive clarity!

**Review by dooart — APPROVED** [review]

**Pull request merged**